imap_processing/hi/l1a/histogram.py

```diff
@@ -0,0 +1,122 @@
+"""Unpack IMAP-Hi histogram data."""
+
+import dataclasses
+
+import numpy as np
+import xarray as xr
+from bitstring import ConstBitStream
+from space_packet_parser.parser import Packet
+
+from imap_processing.cdf.global_attrs import ConstantCoordinates
+from imap_processing.cdf.utils import calc_start_time
+from imap_processing.hi import hi_cdf_attrs
+
+
+def create_dataset(packets: list[Packet]) -> xr.Dataset:
+    """Create dataset for a number of Hi Histogram packets.
+
+    Parameters
+    ----------
+    packets : list[Packet]
+        packet list
+
+    Returns
+    -------
+    xr.dataset
+        dataset with all metadata field data in xr.DataArray
+    """
+    # preallocate the xr.DataArrays for all CDF attributes based on number of packets
+    coords = dict()
+    coords["epoch"] = xr.DataArray(
+        np.empty(len(packets), dtype="datetime64[ns]"),
+        name="epoch",
+        dims=["epoch"],
+        attrs=ConstantCoordinates.EPOCH,
+    )
+    # Histogram data is binned in 90, 4-degree bins
+    # TODO: Confirm whether to define bins by centers or edges. For now centers
+    #    are assumed.
+    coords["angle"] = xr.DataArray(
+        np.arange(2, 360, 4),
+        name="angle",
+        dims=["angle"],
+        attrs=hi_cdf_attrs.hi_hist_l1a_angle_attrs.output(),
+    )
+
+    data_vars = dict()
+    data_vars["esa_step"] = xr.DataArray(
+        np.empty(len(packets), np.dtype("uint8")),
+        dims=["epoch"],
+        attrs=dataclasses.replace(
+            hi_cdf_attrs.hi_hist_l1a_esa_step_attrs,
+            catdesc=packets[0].data["ESA_STEP"].long_description,
+            fieldname="ESA_STEP",
+            label_axis="ESA_STEP",
+        ).output(),
+    )
+
+    data_vars["acq_start"] = xr.DataArray(
+        np.empty(len(packets), np.dtype("uint16")),
+        dims=["epoch"],
+        attrs=dataclasses.replace(
+            hi_cdf_attrs.hi_hist_l1a_acq_start_attrs,
+            catdesc=packets[0].data["ACQ_START"].long_description,
+            fieldname="ACQ_START",
+            label_axis="ACQ_START",
+        ).output(),
+    )
+
+    # This dictionary defines the names and order of the 24, 90 element counter arrays
+    # TODO: This implementation will result in a CDF with variables consisting of the
+    #    counter_def_dict key + tuple string. Need to verify if this is OK with
+    #    instrument team
+    counter_def_dict = {
```

Comment:
I found it hard to parse the nested for loops at first glance, so I am just giving a quick suggestion in case this appeals to you or sparks another idea. But, the way it currently works is fine as well if you like that style better.
```python
qual_list = ["QUAL_" + x for x in ("AB", "C1C2", "AC1", "BC1", "ABC1", "AC1C2", "BC1C2", "ABC1C2")]
long_list = ["LONG_" + ...]
misc_list = ["MISC_" + ...]
for key_counter in qual_list + long_list + misc_list:
    data_vars[key_counter] = ...
```